RecoEcal/EgammaClusterAlgos/interface/SCEnergyCorrectorSemiParm.h

```diff
@@ -5,80 +5,182 @@
+          meanOutTrans_(meanLow, meanHigh),
+          sigmaOutTrans_(sigmaLow, sigmaHigh) {}
+    RegParam(edm::ConsumesCollector cc,
+             std::string meanKey = "",
+             float meanLow = 0,
+             float meanHigh = 0,
+             std::string sigmaKey = "",
+             float sigmaLow = 0,
+             float sigmaHigh = 0)
+        : RegParam(meanKey, meanLow, meanHigh, sigmaKey, sigmaLow, sigmaHigh) {
+      setTokens(cc);
+    }
+    template <edm::Transition esTransition = edm::Transition::BeginLuminosityBlock>
+    void setTokens(edm::ConsumesCollector cc);
+    void setForests(const edm::EventSetup& setup);
 
-  edm::ESHandle<CaloTopology> calotopo_;
-  edm::ESHandle<CaloGeometry> calogeom_;
+    double mean(const std::vector<float>& data) const;
+    double sigma(const std::vector<float>& data) const;
 
-  edm::ESGetToken<CaloTopology, CaloTopologyRecord> tokenCaloTopo_;
-  edm::ESGetToken<CaloGeometry, CaloGeometryRecord> tokenCaloGeom_;
-  edm::ESGetToken<GBRForestD, GBRDWrapperRcd> tokenRegressionKeyEB_;
-  edm::ESGetToken<GBRForestD, GBRDWrapperRcd> tokenUncertaintyKeyEB_;
-  edm::ESGetToken<GBRForestD, GBRDWrapperRcd> tokenRegressionKeyEE_;
-  edm::ESGetToken<GBRForestD, GBRDWrapperRcd> tokenUncertaintyKeyEE_;
+  private:
+    std::string meanKey_;
+    std::string sigmaKey_;
+    EgammaBDTOutputTransformer meanOutTrans_;
+    EgammaBDTOutputTransformer sigmaOutTrans_;
+    edm::ESHandle<GBRForestD> meanForest_;
+    edm::ESHandle<GBRForestD> sigmaForest_;
+    edm::ESGetToken<GBRForestD, GBRDWrapperRcd> meanForestToken_;
+    edm::ESGetToken<GBRForestD, GBRDWrapperRcd> sigmaForestToken_;
+  };
+
+  //barrel = always ecal barrel, endcap may be ECAL or HGCAL
+  RegParam regParamBarrel_;
+  RegParam regParamEndcap_;
+
+  edm::ESHandle<CaloTopology> caloTopo_;
+  edm::ESHandle<CaloGeometry> caloGeom_;
+  edm::ESGetToken<CaloTopology, CaloTopologyRecord> caloTopoToken_;
+  edm::ESGetToken<CaloGeometry, CaloGeometryRecord> caloGeomToken_;
 
   edm::EDGetTokenT<EcalRecHitCollection> tokenEBRecHits_;
   edm::EDGetTokenT<EcalRecHitCollection> tokenEERecHits_;
+  edm::EDGetTokenT<reco::PFRecHitCollection> tokenHgcalRecHits_;
   edm::EDGetTokenT<reco::VertexCollection> tokenVertices_;
 
+  edm::Handle<EcalRecHitCollection> recHitsEB_;
+  edm::Handle<EcalRecHitCollection> recHitsEE_;
+  edm::Handle<reco::PFRecHitCollection> recHitsHgcal_;
   edm::Handle<reco::VertexCollection> vertices_;
-  edm::Handle<EcalRecHitCollection> rechitsEB_;
-  edm::Handle<EcalRecHitCollection> rechitsEE_;
 
   edm::InputTag ecalHitsEBInputTag_;
   edm::InputTag ecalHitsEEInputTag_;
+  edm::InputTag hgcalHitsInputTag_;
+
   edm::InputTag vertexInputTag_;
 
-  std::string regressionKeyEB_;
-  std::string uncertaintyKeyEB_;
-  std::string regressionKeyEE_;
-  std::string uncertaintyKeyEE_;
+  //returns barrel for ecal barrel, otherwise returns endcap
+  const RegParam& getRegParam(const DetId& detId) const {
```

Comment:
will change

Reply:
so here we have the problem ` const auto& regParam = getRegParam(sc.seed()->seed());`

Renaming getRegParam() here is a problem and would make it awkward. Also this is not a pure accessor, it figures out which RegParam to get.  If this was an inlined simple accessor I would agree with you. However given it is not and renaming it would force a rename of regParam it gets , I think its better to leave it the way it was.